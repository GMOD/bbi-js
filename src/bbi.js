import Window from './window'

const LRU = require('quick-lru')
const { Parser } = require('@gmod/binary-parser')
const Long = require('long')
const LocalFile = require('./localFile')

const BIG_WIG_MAGIC = -2003829722
const BIG_BED_MAGIC = -2021002517

export default class BBIFile {
  constructor({
    filehandle,
    path,
    cacheSize,
    fetchSizeLimit,
    chunkSizeLimit,
    renameRefSeqs = n => n,
  }) {
    this.renameRefSeq = renameRefSeqs

    if (filehandle) {
      this.bbi = filehandle
    } else if (path) {
      this.bbi = new LocalFile(path)
    } else {
      throw new Error('no file given')
    }

    this.featureCache = new LRU({
      maxSize: cacheSize !== undefined ? cacheSize : 20000,
      length: featureArray => featureArray.length,
    })

    this.fetchSizeLimit = fetchSizeLimit || 50000000
    this.chunkSizeLimit = chunkSizeLimit || 10000000
    this.gotHeader = this.getHeader()
  }

  // mutates obj for keys ending with '64' to longs, and removes the '64' suffix
  /* eslint no-param-reassign: ["error", { "props": false }] */
  convert64Bits(obj) {
    const keys = Object.keys(obj)
    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i]
      const val = obj[key]
      if (key.endsWith('64')) {
        obj[key.slice(0, -2)] = Long.fromBytes(
          val,
          false,
          !this.isBigEndian,
        ).toNumber()
        delete obj[key]
      } else if (typeof obj[key] === 'object' && val !== null) {
        this.convert64Bits(val)
      }
    }
  }

  async getHeader() {
    if (this.header) {
      return this.header
    }
    const ret = await this.getParsers()
    const buf = Buffer.alloc(2000)
    await this.bbi.read(buf, 0, 2000, 0)
    const res = ret.headerParser.parse(buf)

    const header = res.result
    this.convert64Bits(header)
    this.type = header.magic === BIG_BED_MAGIC ? 'bigbed' : 'bigwig'

    if (header.asOffset) {
      header.autoSql = buf
        .slice(header.asOffset, buf.indexOf(0, header.asOffset))
        .toString('utf8')
    }
    if (header.totalSummaryOffset) {
      const tail = buf.slice(header.totalSummaryOffset)
      header.totalSummary = ret.totalSummaryParser.parse(tail).result
      this.convert64Bits(header.totalSummary)
    }
    const chroms = await this.readChromTree(header)
    Object.assign(header, chroms)
    this.header = header
    return header
  }

  async detectEndianness() {
    const buf = Buffer.allocUnsafe(4)
    await this.bbi.read(buf, 0, 4, 0)
    let ret = buf.readInt32LE(0)
    if (ret === BIG_WIG_MAGIC || ret === BIG_BED_MAGIC) {
      this.isBigEndian = false
      return
    }
    ret = buf.readInt32BE(0)
    if (ret === BIG_WIG_MAGIC || ret === BIG_BED_MAGIC) {
      this.isBigEndian = true
      return
    }
    throw new Error('not a BigWig/BigBed file')
  }

  async getParsers() {
    await this.detectEndianness()

    const le = this.isBigEndian ? 'big' : 'little'
    const headerParser = new Parser()
      .endianess(le)
      .int32('magic')
      .uint16('version')
      .uint16('numZoomLevels')
      .buffer('chromTreeOffset64', { length: 8 })
      .buffer('unzoomedDataOffset64', { length: 8 })
      .buffer('unzoomedIndexOffset64', { length: 8 })
      .uint16('fieldCount')
      .uint16('definedFieldCount')
      .buffer('asOffset64', { length: 8 })
      .buffer('totalSummaryOffset64', { length: 8 })
      .uint32('uncompressBufSize')
      .skip(8) // reserved
      .array('zoomLevels', {
        length: 'numZoomLevels',
        type: new Parser()
          .uint32('reductionLevel')
          .uint32('reserved')
          .buffer('dataOffset64', { length: 8 })
          .buffer('indexOffset64', { length: 8 }),
      })

    const totalSummaryParser = new Parser()
      .endianess(le)
      .buffer('basesCovered64', { length: 8 })
      .double('scoreMin')
      .double('scoreMax')
      .double('scoreSum')
      .double('scoreSumSquares')

    /* istanbul ignore next */
    const chromTreeParser = new Parser()
      .endianess(le)
      .uint32('magic', { assert: s => s === 2026540177 })
      .uint32('blockSize')
      .uint32('keySize')
      .uint32('valSize')
      .buffer('itemCount64', { length: 8 })

    return {
      chromTreeParser,
      totalSummaryParser,
      headerParser,
    }
  }

  async readChromTree(header) {
    const refsByNumber = {}
    const refsByName = {}

    let unzoomedDataOffset = header.unzoomedDataOffset
    while (unzoomedDataOffset % 4 !== 0) {
      unzoomedDataOffset += 1
    }

    const data = Buffer.alloc(unzoomedDataOffset - header.chromTreeOffset)
    await this.bbi.read(
      data,
      0,
      unzoomedDataOffset - header.chromTreeOffset,
      header.chromTreeOffset,
    )

    const p = await this.getParsers()
    const ret = p.chromTreeParser.parse(data).result
    this.convert64Bits(ret)

    const rootNodeOffset = 32
    const bptReadNode = currentOffset => {
      let offset = currentOffset
      if (offset >= data.length) throw new Error('reading beyond end of buffer')
      const isLeafNode = data.readUInt8(offset)
      const cnt = data.readUInt16LE(offset + 2)
      // dlog('ReadNode: ' + offset + '     type=' + isLeafNode + '   count=' + cnt);
      offset += 4
      for (let n = 0; n < cnt; n += 1) {
        if (isLeafNode) {
          // parse leaf node
          let key = ''
          for (let ki = 0; ki < ret.keySize; ki += 1, offset += 1) {
            const charCode = data.readUInt8(offset)
            if (charCode !== 0) {
              key += String.fromCharCode(charCode)
            }
          }
          const refId = data.readUInt32LE(offset)
          const refSize = data.readUInt32LE(offset + 4)
          offset += 8

          const refRec = { name: key, id: refId, length: refSize }

          // dlog(key + ':' + refId + ',' + refSize);
          refsByName[key] = refRec
          refsByNumber[refId] = refRec
        } else {
          // parse index node
          offset += ret.keySize
          let childOffset = Long.fromBytes(offset).toNumber()
          offset += 8
          childOffset -= header.chromTreeOffset
          bptReadNode(childOffset)
        }
      }
    }
    bptReadNode(rootNodeOffset)
    return {
      refsByName,
      refsByNumber,
    }
  }

  getView(scale) {
    const header = this.header
    if (!header.zoomLevels || !header.zoomLevels.length) return null

    if (!this.viewCache || this.viewCache.scale !== scale) {
      this.viewCache = {
        scale,
        view: this.getViewHelper(scale),
      }
    }
    return this.viewCache.view
  }

  getViewHelper(scale) {
    const header = this.header
    const basesPerPx = 1 / scale
    // console.log('getting view for '+basesPerSpan+' bases per span');
    let maxLevel = header.zoomLevels.length
    if (!header.fileSize) {
      // if we don't know the file size, we can't fetch the highest zoom level :-(
      maxLevel -= 1
    }

    for (let i = maxLevel; i > 0; i -= 1) {
      const zh = header.zoomLevels[i]
      if (zh && zh.reductionLevel <= 2 * basesPerPx) {
        const indexLength =
          i < header.zoomLevels.length - 1
            ? header.zoomLevels[i + 1].dataOffset - zh.indexOffset
            : header.fileSize - 4 - zh.indexOffset
        // console.log( 'using zoom level '+i);
        return new Window(this, zh.indexOffset, indexLength, true)
      }
    }
    return this.getUnzoomedView()
  }

  getUnzoomedView() {
    const header = this.header
    if (!this.unzoomedView) {
      let cirLen = 4000
      const nzl = header.zoomLevels[0]
      if (nzl) {
        cirLen = nzl.dataOffset - header.unzoomedIndexOffset
      }
      this.unzoomedView = new Window(
        this,
        header.unzoomedIndexOffset,
        cirLen,
        false,
        this.autoSql,
      )
    }
    return this.unzoomedView
  }
}
