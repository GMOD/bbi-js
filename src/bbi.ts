import { Parser } from '@gmod/binary-parser'
import { LocalFile, RemoteFile, GenericFilehandle } from 'generic-filehandle'
import { Observable, Observer } from 'rxjs'
import { reduce } from 'rxjs/operators'
import AbortablePromiseCache from 'abortable-promise-cache'
import QuickLRU from 'quick-lru'
import { BlockView } from './blockView'

const BIG_WIG_MAGIC = -2003829722
const BIG_BED_MAGIC = -2021002517

export interface Feature {
  start: number
  end: number
  score: number
  rest?: string // for bigbed line
  minScore?: number // for summary line
  maxScore?: number // for summary line
  summary?: boolean // is summary line
  uniqueId?: string // for bigbed contains uniqueId calculated from file offset
  field?: number // used in bigbed searching
}
interface Statistics {
  scoreSum: number
  basesCovered: number
  scoreSumSquares: number
}

interface RefInfo {
  name: string
  id: number
  length: number
}
export interface Header {
  autoSql: string
  totalSummary: Statistics
  zoomLevels: any
  unzoomedIndexOffset: number
  unzoomedDataOffset: number
  definedFieldCount: number
  uncompressBufSize: number
  chromTreeOffset: number
  fileSize: number
  extHeaderOffset: number
  isBigEndian: boolean
  fileType: string
  refsByName: { [key: string]: number }
  refsByNumber: { [key: number]: RefInfo }
}

/* get the compiled parsers for different sections of the bigwig file
 *
 * @param isBE - is big endian, typically false
 * @return an object with compiled parsers
 */
function getParsers(isBE: boolean): any {
  const le = isBE ? 'big' : 'little'
  const headerParser = new Parser()
    .endianess(le)
    .int32('magic')
    .uint16('version')
    .uint16('numZoomLevels')
    .uint64('chromTreeOffset')
    .uint64('unzoomedDataOffset')
    .uint64('unzoomedIndexOffset')
    .uint16('fieldCount')
    .uint16('definedFieldCount')
    .uint64('asOffset') // autoSql offset, used in bigbed
    .uint64('totalSummaryOffset')
    .uint32('uncompressBufSize')
    .uint64('extHeaderOffset') // name index offset, used in bigbed
    .array('zoomLevels', {
      length: 'numZoomLevels',
      type: new Parser()
        .uint32('reductionLevel')
        .uint32('reserved')
        .uint64('dataOffset')
        .uint64('indexOffset'),
    })

  const totalSummaryParser = new Parser()
    .endianess(le)
    .uint64('basesCovered')
    .double('scoreMin')
    .double('scoreMax')
    .double('scoreSum')
    .double('scoreSumSquares')

  const chromTreeParser = new Parser()
    .endianess(le)
    .uint32('magic')
    .uint32('blockSize')
    .uint32('keySize')
    .uint32('valSize')
    .uint64('itemCount')

  const isLeafNode = new Parser()
    .endianess(le)
    .uint8('isLeafNode')
    .skip(1)
    .uint16('cnt')

  return {
    chromTreeParser,
    totalSummaryParser,
    headerParser,
    isLeafNode,
  }
}

export interface RequestOptions {
  signal?: AbortSignal
  headers?: Record<string, string>
  [key: string]: unknown
}

export abstract class BBI {
  protected bbi: GenericFilehandle

  protected headerCache = new AbortablePromiseCache({
    cache: new QuickLRU({ maxSize: 1 }),
    fill: async (params: any, signal?: AbortSignal) => {
      return this._getHeader({ ...params, signal })
    },
  })

  protected renameRefSeqs: (a: string) => string

  /* fetch and parse header information from a bigwig or bigbed file
   * @param abortSignal - abort the operation, can be null
   * @return a Header object
   */
  public getHeader(opts: RequestOptions | AbortSignal = {}) {
    const options = 'aborted' in opts ? { signal: opts } : opts
    return this.headerCache.get(
      JSON.stringify(options),
      options,
      options.signal,
    )
  }

  /*
   * @param filehandle - a filehandle from generic-filehandle or implementing something similar to the node10 fs.promises API
   * @param path - a Local file path as a string
   * @param url - a URL string
   * @param renameRefSeqs - an optional method to rename the internal reference sequences using a mapping function
   */
  public constructor(
    options: {
      filehandle?: GenericFilehandle
      path?: string
      url?: string
      renameRefSeqs?: (a: string) => string
    } = {},
  ) {
    const { filehandle, renameRefSeqs, path, url } = options
    this.renameRefSeqs = renameRefSeqs || ((s: string): string => s)
    if (filehandle) {
      this.bbi = filehandle
    } else if (url) {
      this.bbi = new RemoteFile(url)
    } else if (path) {
      this.bbi = new LocalFile(path)
    } else {
      throw new Error('no file given')
    }
  }

  private async _getHeader(opts: RequestOptions) {
    const header = await this._getMainHeader(opts)
    const chroms = await this._readChromTree(header, opts)
    return { ...header, ...chroms }
  }

  private async _getMainHeader(
    opts: RequestOptions,
    requestSize = 2000,
  ): Promise<Header> {
    const { buffer } = await this.bbi.read(
      Buffer.alloc(requestSize),
      0,
      requestSize,
      0,
      opts,
    )
    const isBigEndian = this._isBigEndian(buffer)
    const ret = getParsers(isBigEndian)
    const header = ret.headerParser.parse(buffer).result
    header.fileType = header.magic === BIG_BED_MAGIC ? 'bigbed' : 'bigwig'
    if (
      header.asOffset > requestSize ||
      header.totalSummaryOffset > requestSize
    ) {
      return this._getMainHeader(opts, requestSize * 2)
    }
    if (header.asOffset) {
      header.autoSql = buffer
        .slice(header.asOffset, buffer.indexOf(0, header.asOffset))
        .toString('utf8')
    }
    if (header.totalSummaryOffset > requestSize) {
      return this._getMainHeader(opts, requestSize * 2)
    }
    if (header.totalSummaryOffset) {
      const tail = buffer.slice(header.totalSummaryOffset)
      header.totalSummary = ret.totalSummaryParser.parse(tail).result
    }
    return { ...header, isBigEndian }
  }

  private _isBigEndian(buffer: Buffer): boolean {
    let ret = buffer.readInt32LE(0)
    if (ret === BIG_WIG_MAGIC || ret === BIG_BED_MAGIC) {
      return false
    }
    ret = buffer.readInt32BE(0)
    if (ret === BIG_WIG_MAGIC || ret === BIG_BED_MAGIC) {
      return true
    }
    throw new Error('not a BigWig/BigBed file')
  }

  // todo: add progress if long running
  private async _readChromTree(header: Header, opts: { signal?: AbortSignal }) {
    const isBE = header.isBigEndian
    const le = isBE ? 'big' : 'little'
    const refsByNumber: {
      [key: number]: { name: string; id: number; length: number }
    } = []
    const refsByName: { [key: string]: number } = {}
    const { chromTreeOffset } = header
    let { unzoomedDataOffset } = header

    while (unzoomedDataOffset % 4 !== 0) {
      unzoomedDataOffset += 1
    }

    const { buffer: data } = await this.bbi.read(
      Buffer.alloc(unzoomedDataOffset - chromTreeOffset),
      0,
      unzoomedDataOffset - chromTreeOffset,
      chromTreeOffset,
      opts,
    )

    const p = getParsers(isBE)
    const { keySize } = p.chromTreeParser.parse(data).result
    const leafNodeParser = new Parser()
      .endianess(le)
      .string('key', { stripNull: true, length: keySize })
      .uint32('refId')
      .uint32('refSize')
    const nonleafNodeParser = new Parser()
      .endianess(le)
      .skip(keySize)
      .uint64('childOffset')
    const rootNodeOffset = 32
    const bptReadNode = async (currentOffset: number): Promise<void> => {
      let offset = currentOffset
      if (offset >= data.length) {
        throw new Error('reading beyond end of buffer')
      }
      const ret = p.isLeafNode.parse(data.slice(offset))
      const { isLeafNode, cnt } = ret.result
      offset += ret.offset
      if (isLeafNode) {
        for (let n = 0; n < cnt; n += 1) {
          const leafRet = leafNodeParser.parse(data.slice(offset))
          offset += leafRet.offset
          const { key, refId, refSize } = leafRet.result
          const refRec = { name: key, id: refId, length: refSize }
          refsByName[this.renameRefSeqs(key)] = refId
          refsByNumber[refId] = refRec
        }
      } else {
        // parse index node
        const nextNodes = []
        for (let n = 0; n < cnt; n += 1) {
          const nonleafRet = nonleafNodeParser.parse(data.slice(offset))
          let { childOffset } = nonleafRet.result
          offset += nonleafRet.offset
          childOffset -= chromTreeOffset
          nextNodes.push(bptReadNode(childOffset))
        }
        await Promise.all(nextNodes)
      }
    }
    await bptReadNode(rootNodeOffset)
    return {
      refsByName,
      refsByNumber,
    }
  }

  /*
   * fetches the "unzoomed" view of the bigwig data. this is the default for bigbed
   * @param abortSignal - a signal to optionally abort this operation
   */
  protected async getUnzoomedView(opts: RequestOptions): Promise<BlockView> {
    const {
      unzoomedIndexOffset,
      zoomLevels,
      refsByName,
      uncompressBufSize,
      isBigEndian,
      fileType,
    } = await this.getHeader(opts)
    const nzl = zoomLevels[0]
    const cirLen = nzl ? nzl.dataOffset - unzoomedIndexOffset : 4000
    return new BlockView(
      this.bbi,
      refsByName,
      unzoomedIndexOffset,
      cirLen,
      isBigEndian,
      uncompressBufSize > 0,
      fileType,
    )
  }

  /*
   * abstract method - get the view for a given scale
   */
  protected abstract getView(
    scale: number,
    opts: RequestOptions,
  ): Promise<BlockView>

  /**
   * Gets features from a BigWig file
   *
   * @param refName - The chromosome name
   * @param start - The start of a region
   * @param end - The end of a region
   * @param opts - An object containing basesPerSpan (e.g. pixels per basepair) or scale used to infer the zoomLevel to use
   */
  public async getFeatureStream(
    refName: string,
    start: number,
    end: number,
    opts: RequestOptions & { scale?: number; basesPerSpan?: number } = {
      scale: 1,
    },
  ): Promise<Observable<Feature[]>> {
    await this.getHeader(opts)
    const chrName = this.renameRefSeqs(refName)
    let view: BlockView

    if (opts.basesPerSpan) {
      view = await this.getView(1 / opts.basesPerSpan, opts)
    } else if (opts.scale) {
      view = await this.getView(opts.scale, opts)
    } else {
      view = await this.getView(1, opts)
    }

    if (!view) {
      throw new Error('unable to get block view for data')
    }
    return new Observable((observer: Observer<Feature[]>): void => {
      view.readWigData(chrName, start, end, observer, opts)
    })
  }

  public async getFeatures(
    refName: string,
    start: number,
    end: number,
    opts: RequestOptions & { scale?: number; basesPerSpan?: number } = {
      scale: 1,
    },
  ): Promise<Feature[]> {
    const ob = await this.getFeatureStream(refName, start, end, opts)

    const ret = await ob
      .pipe(reduce((acc, curr) => acc.concat(curr)))
      .toPromise()
    return ret || []
  }
}
