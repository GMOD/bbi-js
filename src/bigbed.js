const BBI = require('./bbi')

export default class BigBed extends BBI {
  getFeatures(refName, start, end, opts = {}) {
    const chrName = this.renameRefSeq(refName)
    let view

    view = this.getView(1)
    if (!view) {
      return null
    }

    return view.readWigData(chrName, start, end)
  }

  parseBigBedBlock(bytes, startOffset) {
    const data = this.window.bwg.newDataView(bytes, startOffset)

    let offset = 0
    while (offset < bytes.byteLength) {
      const chromId = data.getUint32(offset)
      const start = data.getInt32(offset + 4)
      const end = data.getInt32(offset + 8)
      offset += 12
      if (chromId !== this.chr) {
        console.warn('BigBed block is out of current range')
        return
      }

      let rest = ''
      while (offset < bytes.byteLength) {
        const ch = data.getUint8(offset)
        offset += 1
        if (ch !== 0) {
          rest += String.fromCharCode(ch)
        } else {
          break
        }
      }

      const featureData = this.parseBedText(start, end, rest)
      featureData.id = `bb-${startOffset + offset}`
      this.maybeCreateFeature(start, end, featureData)
    }
  }


}
