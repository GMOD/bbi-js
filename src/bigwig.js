const BBI = require('./bbi')

export default class BigWig extends BBI {
  getFeatures(refName, start, end, opts = {}) {
    const chrName = this.renameRefSeq(refName)
    let view

    if (opts.basesPerSpan) {
      view = this.getView(1 / opts.basesPerSpan)
    } else if (opts.scale) {
      view = this.getView(opts.scale)
    } else {
      view = this.getView(1)
    }

    if (!view) {
      return null
    }

    return view.readWigData(chrName, start, end)
  }
}
