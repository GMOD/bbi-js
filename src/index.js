import BBI from './bbi'
import BigWig from './bigwig'
import BigBed from './bigbed'

export { BBI, BigWig, BigBed }
