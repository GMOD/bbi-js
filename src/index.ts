export { BigWig } from './bigwig'
export { BigBed } from './bigbed'
export { Feature, Header, RequestOptions } from './bbi'
