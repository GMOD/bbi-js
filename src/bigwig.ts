import { BlockView } from './blockView'
import { BBI, RequestOptions } from './bbi'

export class BigWig extends BBI {
  /**
   * Retrieves a BlockView of a specific zoomLevel
   *
   * @param refName - The chromosome name
   * @param start - The start of a region
   * @param end - The end of a region
   * @param opts - An object containing basesPerSpan (e.g. pixels per basepair) or scale used to infer the zoomLevel to use
   */
  protected async getView(
    scale: number,
    opts: RequestOptions,
  ): Promise<BlockView> {
    const { zoomLevels, refsByName, fileSize, isBigEndian, uncompressBufSize } =
      await this.getHeader(opts)
    const basesPerPx = 1 / scale
    let maxLevel = zoomLevels.length
    if (!fileSize) {
      // if we don't know the file size, we can't fetch the highest zoom level :-(
      maxLevel -= 1
    }

    for (let i = maxLevel; i >= 0; i -= 1) {
      const zh = zoomLevels[i]
      if (zh && zh.reductionLevel <= 2 * basesPerPx) {
        const indexLength =
          i < zoomLevels.length - 1
            ? zoomLevels[i + 1].dataOffset - zh.indexOffset
            : fileSize - 4 - zh.indexOffset
        return new BlockView(
          this.bbi,
          refsByName,
          zh.indexOffset,
          indexLength,
          isBigEndian,
          uncompressBufSize > 0,
          'summary',
        )
      }
    }
    return this.getUnzoomedView(opts)
  }
}
