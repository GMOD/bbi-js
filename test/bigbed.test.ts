/* eslint @typescript-eslint/explicit-function-return-type: 0 */
import BED from '@gmod/bed'
import { BigBed } from '../src/'
import { LocalFile } from 'generic-filehandle'

describe('bigbed formats', () => {
  it('loads small bigbed file', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/hg18.bb'),
    })
    const { autoSql } = await ti.getHeader()
    const feats = await ti.getFeatures('chr7', 0, 100000)
    const parser = new BED({ autoSql })
    const lines = feats.map(feat => {
      const { start, end, rest, uniqueId } = feat
      return parser.parseLine(`chr7\t${start}\t${end}\t${rest}`, { uniqueId })
    })
    expect(lines.slice(0, 5)).toMatchSnapshot()
  })
  it('loads volvox.bb', async () => {
    const filehandle = new LocalFile(require.resolve('./data/volvox.bb'))
    const ti = new BigBed({
      filehandle,
    })
    const spy = jest.spyOn(filehandle, 'read')
    const feats = await ti.getFeatures('chrA', 0, 160)
    expect(spy.mock.calls.length).toBeLessThanOrEqual(3)
    expect(feats).toEqual([])
  })
  it('searchExtraIdex returns null on file with no extra index', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/volvox.bb'),
    })
    await ti.readIndices()
    const res = await ti.searchExtraIndex('EDEN.1')
    expect(res).toMatchSnapshot()
  })

  it('searchExtraIndex on file with no name index', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/chr22.bb'),
    })
    await ti.readIndices()
    const res = await ti.searchExtraIndex('ENST00000467796.2')
    expect(res).toEqual([])
  })

  it('searchExtraIndex name in gencode', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/chr22_with_name_index.bb'),
    })
    await ti.readIndices()
    const res = await ti.searchExtraIndex('ENST00000467796.2')
    expect(res).toMatchSnapshot()
  })

  it('searchExtraIndex in bigbed with multiple extra indexes on the gene name index', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/chr22_with_name_and_geneName_index.bb'),
    })
    await ti.readIndices()
    const res2 = await ti.searchExtraIndex('SYCE3')
    expect(res2).toMatchSnapshot()
  })
  it('2057 contigs', async () => {
    const ti = new BigBed({
      path: require.resolve('./data/2057.bb'),
    })
    const header = await ti.getHeader()
    expect(Object.keys(header.refsByName).length).toEqual(2057)
  })
  it('bigbed file with large header', async () => {
    const filehandle = new LocalFile(require.resolve('./data/clinvarCnv.bb'))
    const ti = new BigBed({
      filehandle,
    })
    const spy = jest.spyOn(filehandle, 'read')
    const header = await ti.getHeader()
    expect(spy.mock.calls.length).toBeLessThanOrEqual(3)
    expect(header).toBeTruthy()
  })
  it('bigbed file consistent file ID', async () => {
    const filehandle = new LocalFile(require.resolve('./data/clinvarCnv.bb'))
    const ti = new BigBed({
      filehandle,
    })
    const feats1 = await ti.getFeatures('chr21', 8850000, 10050000)
    const feats2 = await ti.getFeatures('chr21', 9550000, 10750000)
    const f1 = feats1.find(f => f.start === 9734795)
    const f2 = feats2.find(f => f.start === 9734795)
    expect(f1?.uniqueId).toEqual(f2?.uniqueId)
    const f11 = feats1.filter(f => f.start === 9734795)
    const f22 = feats2.filter(f => f.start === 9734795)
    expect(f11).toEqual(f22)
  })
})
