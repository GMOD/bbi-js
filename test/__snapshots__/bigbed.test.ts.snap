// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bigbed formats loads small bigbed file 1`] = `
Array [
  Object {
    "chrom": "chr7",
    "chromEnd": 73584,
    "chromStart": 54028,
    "name": "uc003sii.2",
    "reserved": "255,0,0",
    "score": 0,
    "spID": "AL137655",
    "strand": -1,
    "thickEnd": 54028,
    "thickStart": 54028,
    "uniqueId": "bb-0-0",
  },
  Object {
    "chrom": "chr7",
    "chromEnd": 61569,
    "chromStart": 60328,
    "name": "uc010krx.1",
    "reserved": "255,0,0",
    "score": 0,
    "spID": "PDGFA",
    "strand": -1,
    "thickEnd": 60328,
    "thickStart": 60328,
    "uniqueId": "bb-0-1",
  },
  Object {
    "chrom": "chr7",
    "chromEnd": 63529,
    "chromStart": 62967,
    "name": "uc003sij.2",
    "reserved": "255,0,0",
    "score": 0,
    "spID": "DQ576410",
    "strand": -1,
    "thickEnd": 63366,
    "thickStart": 62967,
    "uniqueId": "bb-0-2",
  },
  Object {
    "chrom": "chr7",
    "chromEnd": 64107,
    "chromStart": 64068,
    "name": "uc003sil.1",
    "reserved": "255,0,0",
    "score": 0,
    "spID": "DQ584609",
    "strand": -1,
    "thickEnd": 64068,
    "thickStart": 64068,
    "uniqueId": "bb-0-3",
  },
  Object {
    "chrom": "chr7",
    "chromEnd": 65220,
    "chromStart": 65159,
    "name": "uc003sim.1",
    "reserved": "255,0,0",
    "score": 0,
    "spID": "DQ600587",
    "strand": -1,
    "thickEnd": 65159,
    "thickStart": 65159,
    "uniqueId": "bb-0-4",
  },
]
`;

exports[`bigbed formats searchExtraIdex returns null on file with no extra index 1`] = `
Array [
  Object {
    "chromId": 0,
    "end": 9000,
    "field": 3,
    "rest": "EDEN.1	1000	+	1200	7608	0,0,0	4	451,903,501,2001	0,1950,3950,5950	EDEN	EDEN	-	-	-",
    "start": 1049,
    "uniqueId": "bb-0-0",
  },
]
`;

exports[`bigbed formats searchExtraIndex in bigbed with multiple extra indexes on the gene name index 1`] = `
Array [
  Object {
    "chromId": 0,
    "end": 50562905,
    "field": 13,
    "rest": "ENST00000406915.3	1000	-	50551247	50556405	0,0,0	3	291,109,48	0,5185,11746	ENSG00000217442.3	SYCE3	protein_coding	protein_coding	basic,appris_principal,exp_conf,CCDS",
    "start": 50551111,
    "uniqueId": "bb-0-221",
  },
  Object {
    "chromId": 0,
    "end": 50556432,
    "field": 13,
    "rest": "ENST00000402753.1	1000	-	50551247	50556405	0,0,0	2	206,136	0,5100	ENSG00000217442.3	SYCE3	protein_coding	protein_coding	basic,appris_principal,CCDS",
    "start": 50551196,
    "uniqueId": "bb-0-222",
  },
]
`;

exports[`bigbed formats searchExtraIndex name in gencode 1`] = `
Array [
  Object {
    "chromId": 0,
    "end": 50799305,
    "field": 3,
    "rest": "ENST00000467796.2	1000	+	0	0	0,0,0	3	41,121,651	0,1919,15401	ENSG00000184319.12	RPL23AP82	processed_transcript	transcribed_unprocessed_pseudogene	-",
    "start": 50783253,
    "uniqueId": "bb-0-291",
  },
]
`;
